Currency: adds more query support

adityatandon007 opened the pull request:
**Currency Conversion IA**: This PR adds support regarding currency conversion triggering

## Description of new Instant Answer, or changes
Adds support for Currency IA

## Related Issues and Discussions
Fixes #3254 

## People to notify
@moollaza @pjhampton 

<!-- DO NOT REMOVE -->
---

<!-- The Instant Answer ID can be found by clicking the `?` icon beside the Instant Answer result on DuckDuckGo.com -->
Instant Answer Page: https://duck.co/ia/view/currency
<!-- FILL THIS IN:                           ^^^^ -->

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

pjhampton commented:
Thanks @adityatandon007. Can you please follow the DDH convention for naming your PRs in the future?

@moollaza please deploy to beta 👍

moollaza commented:
Deloying to Beta 👍

moollaza commented:
@adityatandon007 let's add support for words like "currency converter", "currency conversion", "calculator", "calculate", etc.

E.g.
- "usd to cad currency (conversion|converter)"
- "calculate 1 cad to usd"
- "4 cad to use currency calculator"

You can use this:
```perl
my $keyword_qr = qr/(?:(?<currencyKeyword>(currency ?)?(?:value|price|conversion|converter|calculator|(exchange|conversion)\srate|(value|price)\s?of)?)\s?)/i;
```

We also don't need the optional space at the beginning as we allow for that in the `guard`

Please add tests for those as well 👍

adityatandon007 commented:
@moollaza we need that space we are not allowing it in the `guard` 
[]( zeroclickinfo-spice/lib/DDG/Spice/Currency.pm#L38)

adityatandon007 commented:
@moollaza for now I have removed the space from the keyword_qr and added that in the guard before the last $keyword_qr and support for calculator has been added in the keyword_qr regex.

moollaza commented:
This is looking good @adityatandon007!

> so, I tried my regex though it doesn't looks good (more repetitive), but it doesn't fail earlier cases. Let me know if you have a workaround for those cases or if you think mine version of regex is fine.

I think this works...

```perl
qr/(?:(?<currencyKeyword>(?:((currency|value|price)( (conver(sion|ter)|calculator))?|(conver(sion|ter)|calculator)|(exchange|conversion) rate)|(value|price) of) ?)/i;
```

> One more thing this PR triggers for queries like usd conversion, usd currency. you were saying we should not trigger for those cases. Should I adjust the query so it doesn't trigger for those cases.

Those queries _**should**_ trigger it. They're perfectly valid. Google shows their converter for them as well 👍

adityatandon007 commented:
@moolllaza thanks for the tip but you left an `)` so I got an unmatched error but I resolved the error and your query version is working fine. 

One more thing can you please deploy this to beta.

moollaza commented:
@adityatandon007 this is looking good! I just merged a Currency PR though so there's a conflict here. Can you please resolve? Then we can deploy to Beta 👍

adityatandon007 commented:
@moollaza I tried resolving the conflict but I am getting errors while testing the file after resolving. These are the failing test: 

```#   Failed test 'Testing query usd conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/usd'

#   Failed test 'Testing query convert currency 5k usd to inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/5000/usd/inr'

#   Failed test 'Testing query usd to cad currency converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/cad'

#   Failed test 'Testing query usd to cad currency conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/1/usd/cad'

#   Failed test 'Testing query 4 cad to usd currency calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 53.
#     Structures begin differing at:
#          $got->{call} = '/js/spice/currency/1/usd/eur'
#     $expected->{call} = '/js/spice/currency/4/cad/usd'
# Looks like you failed 10 tests of 120.
```
I tried to find the reason and came to this conclusion that somebody changed the results of the earlier queries and it is now defaulting to `usd` to `eur` for those test cases. This is happening because of this piece of code 
```handle query_lc => sub {

    # returns for plain language queries such as 'currency converter'
    if(m/$lang_qr/) {
        my $from = getLocalCurrency();
        my $to = 'usd';

        if($from eq $to) {
            $to = 'eur';
        }

        return checkCurrencyCode(1, $from, $to);
    }
    ```
I don't know why we are defaulting to these. If I remove this piece of code it works fine, ie falls back to our earlier responses. One more thing some of our test cases were deleted too from test files, I don't know what to do? Should I simply go and replace my changes with the new ones (to just get rid of the conflict), if it is so then the aim which this PR is supposed to do (query support) would be simply lost. 

P.S. I want your suggestions on these.

pjhampton commented:
It seems `if(m/$lang_qr/)` is not strict enough. You are going to have to tighten that regex so it has a full match, rather than a partial match. This was an oversight on my behalf but we can't merge this until it's fixed.

adityatandon007 commented:
@pjhampton If I use your `$keyword_qr` and `$guard` of course with `$lang_qr` it is failing these `17`  test: 
```
#   Failed test 'Expected result but dont get one on us dollar conversion rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on bitcoin converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on value of inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on price of bitcoin'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to inr exchange rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on inr to cny conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 500 cny to usd conversion rate'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 5k usd to yuan converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on convert currency 5k usd to inr'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.
#   Failed test 'Expected result but dont get one on conversion 5k usd to yuan'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on conversion 5k usd to uk'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad currency converter'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad currency conversion'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on 4 cad to usd currency calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.

#   Failed test 'Expected result but dont get one on usd to cad calculator'
#   at /home/codio/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/DDG/Test/Spice.pm line 56.
# Looks like you failed 17 tests of 120.
```
Any suggestions on how to resolve the conflict would be helpful?

pjhampton commented:
Are these your new tests?

pjhampton commented on share/spice/currency/currencyNames.txt:
```diff
@@ -44,7 +44,7 @@ etb,ethiopian birr,ethiopia birr,
 fjd,fijian dollar,fiji dollar,
 fkp,falkland island pound,
-gbp,british pound,united kingdom pound,uk pound,uk pounds,pound,sterling,pound sterling,
+gbp,british pound,united kingdom pound,uk pound,uk pounds,pound,sterling,pound sterling,uk
```
Let's remove `uk`. This makes no sense

adityatandon007 commented:
@pjhampton Please review

pjhampton commented:
No problem. Thanks for this @adityatandon007! I will review today. 
Is there anything else you need from me to get this past the finish line?

adityatandon007 commented:
@pjhampton I am finished with this. I think only @moollaza has the right to deploy it to beta. Can you deploy it to beta as well. I w'd love to see this live. It would be very kind of you it gets deployed today?

pjhampton commented:
@adityatandon007 @moollaza is a while for a while. He should be back soon. 
I will see what I can do as I want this pushed through too :+1:

pjhampton commented on t/Currency.t:
```diff
@@ -425,6 +425,12 @@ ddg_spice_test(
         caller => 'DDG::Spice::Currency',
         is_cached => 0
     ),
+				'currency calculator' => test_spice(
```
oops... let's move this back

pjhampton replied:
I can sort in my pr tomorrow 👍

pjhampton commented:
For anybody's reference. We hit this with the internal query tool and all was working great. Just a minor formatting issue which I will fix up tomorrow.

Thanks for your hard work as usual @adityatandon007 💯  🏅 ✨

bsstoner commented:
This was released to production 🚀
